capacity4more/modules/c4m/features_og/c4m_features_og_manage/c4m_manage_categories.inc

```diff
@@ -0,0 +1,834 @@
+    $vocabulary->hierarchy = $hierarchy;
+    taxonomy_vocabulary_save($vocabulary);
+  }
+  drupal_set_message(t('The configuration options have been saved.'));
+}
+
+/**
+ * Form function for the category type edit form.
+ *
+ * To avoid duplicated code, using the term edit form to build this form,
+ * as it's almost the same form, Just unset the parent fields and change the title and description.
+ *
+ * @ingroup forms
+ * @see taxonomy_form_term()
+ */
+function c4m_features_og_manage_categories_type_form($form, &$form_state, $tid) {
+  $form = c4m_features_og_manage_categories_term_form($form, $form_state, $tid);
+  unset($form['title']);
+  unset($form['description']);
+  unset($form['parent']);
+
+  $title['element'] = array(
+    '#tag' => 'h2',
+    '#attributes' => array(
+      'class' => array('type-edit-title-text'),
+    ),
+    '#value' => t('Manage types of category'),
+  );
+
+  $description['element'] = array(
+    '#tag' => 'h4',
+    '#attributes' => array(
+      'class' => array('type-edit-description-text'),
+    ),
+    '#value' => t('Change the types of category of this Group.'),
+  );
+
+  $form['title'] = array(
+    '#markup' => theme_html_tag($title),
+    '#weight' => -10,
+  );
+
+  $form['description'] = array(
+    '#markup' => theme_html_tag($description),
+    '#weight' => -5,
+  );
+  return $form;
+}
+
+/**
+ * Validation handler for the custom type form.
+ *
+ * @see c4m_features_og_manage_categories_type_form()
+ */
+function c4m_features_og_manage_categories_type_form_validate($form, &$form_state) {
+  entity_form_field_validate('taxonomy_term', $form, $form_state);
+}
+
+/**
+ * Submit handler to update a type.
+ *
+ * @see c4m_features_og_manage_categories_type_form()
+ */
+function c4m_features_og_manage_categories_type_form_submit($form, &$form_state) {
+  if ($form_state['triggering_element']['#value'] == t('Delete')) {
+    // Execute the term deletion.
+    if ($form_state['values']['delete'] === TRUE) {
+      c4m_features_og_manage_taxonomy_term_confirm_delete_submit($form, $form_state);
+      return;
+    }
+    // Rebuild the form to confirm term deletion.
+    $form_state['rebuild'] = TRUE;
+    $form_state['confirm_delete'] = TRUE;
+    return;
+  }
+  if ($form_state['triggering_element']['#value'] == t('Cancel')) {
+    return;
+  }
+
+  $term = c4m_features_og_manage_taxonomy_form_term_submit_build_taxonomy_term($form, $form_state);
+
+  taxonomy_term_save($term);
+  drupal_set_message(t('Updated category type %term.', array('%term' => $term->name)));
+  watchdog('taxonomy', 'Updated category type %term.', array('%term' => $term->name), WATCHDOG_NOTICE, l(t('edit'), 'manage/categories/type/' . $term->tid . '/edit'));
+  // Clear the page and block caches to avoid stale data.
+  cache_clear_all();
+}
+
+/**
+ * Form function for the term edit form.
+ *
+ * @ingroup forms
+ * @see taxonomy_form_term()
+ */
+function c4m_features_og_manage_categories_term_form($form, &$form_state, $tid) {
```

Comment:
This function is handling to much. 1 function = 1 functionality. Keep the "controllers" simple.

Use separate paths, callbacks & submit handlers to keep the code clean. Put common functionality in separate helper method.